source/v1/I2CDetail.cpp

```diff
@@ -0,0 +1,258 @@
+        // TODO: This could be lock-free
+        // Enter a critical section
+        mbed::util::CriticalSectionLock lock;
+        // Advance to the next transaction
+        TransactionQueue = t->next;
+        if (TransactionQueue) {
+            // Initiate the next transaction
+            start_transaction();
+        } else {
+            power_down();
+        }
+    } else {
+        // TODO: this could be lock-free
+        mbed::util::CriticalSectionLock lock;
+        t->current = t->current->get_next();
+        start_segment();
+    }
+}
+void I2CResourceManager::event_cb(I2CTransaction * t, int event) {
+    // TODO: Should these be else-ifs?
+    if (event & I2C_EVENT_ERROR_NO_SLAVE) {
+        t->_noslaveCB(t, event);
+    }
+    if (event & I2C_EVENT_TRANSFER_EARLY_NACK) {
+        t->_nakCB(t, event);
+    }
+    if (event & I2C_EVENT_ERROR) {
+        t->_errorCB(t, event);
+    }
+    if (event & I2C_EVENT_TRANSFER_COMPLETE) {
+        t->_doneCB(t, event);
+    }
+    t->issuer->free(t);
+}
+
+
+I2CResourceManager::I2CResourceManager() : TransactionQueue(nullptr) {}
+I2CResourceManager::~I2CResourceManager() {
+    while (TransactionQueue) {
+        I2CTransaction * tx = (I2CTransaction *)(TransactionQueue);
+        TransactionQueue = tx->next;
+        tx->issuer->free(tx);
+    }
+}
+
+class HWI2CResourceManager : public I2CResourceManager {
+public:
+    HWI2CResourceManager(const size_t id, void(*handler)(void)):
+        _i2c(),
+        _id(id),
+        _usage(DMA_USAGE_NEVER),
+        _inited(false),
+        _handler(handler)
+    {
+    }
+    int init(PinName sda, PinName scl) {
+        // calling init during a transaction could cause communication artifacts
+        if (!_inited) {
+            // The init function also set the frequency to 100000
+            i2c_init(&_i2c, sda, scl);
+            _inited = true;
+        } else {
+            // Only support an I2C master on a single pair of pins
+            if (_scl != scl || _sda != sda) {
+                return -1;
+            }
+        }
+        return 0;
+    }
+    int start_segment () {
+        I2CTransaction * t = (I2CTransaction *)TransactionQueue;
+        if (!t) {
+            return -1;
+        }
+        I2CSegment * s = t->current;
+        if (!s) {
+            return -1;
+        }
+        bool stop = (s->get_next() == nullptr) && (!t->repeated);
+        if (s->get_dir() == I2CDirection::Transmit) {
+            i2c_transfer_asynch(&_i2c, s->get_buf(), s->get_len(), nullptr, 0, t->address,
+                stop, (uint32_t)_handler, I2C_EVENT_ALL, _usage);
+        } else {
+            i2c_transfer_asynch(&_i2c, nullptr, 0, s->get_buf(), s->get_len(), t->address,
+                stop, (uint32_t)_handler, I2C_EVENT_ALL, _usage);
+        }
+        return 0;
+    }
+    int start_transaction() {
+        // TODO: This needs a higher-level start/stop lock
```

Comment:
In the absence of such a higher level start/stop lock, presumably this code is broken?  In which case the code needs fixing, otherwise the comment shouldn't be a TODO.